aiida/engine/processes/workchains/workchain.py

```diff
@@ -121,24 +121,53 @@ def on_run(self):
         super().on_run()
         self.node.set_stepper_state_info(str(self._stepper))
 
+    def _resolve_nested_context(self, key: str) -> Tuple[AttributeDict, str]:
+        """
+        Returns a reference to a sub-dictionary of the context and the last key,
+        after resolving a potentially segmented key where required sub-dictionaries are created as needed.
+
+        :param key: A key into the context, where words before a dot are interpreted as a key for a sub-dictionary
+        """
+        ctx = self.ctx
+        ctx_path = key.split('.')
+
+        for path in ctx_path[:-1]:
+            try:
+                ctx = ctx[path]
+            except TypeError:
+                raise AssertionError(
+                    f'The sub-dictionary part `{path}` of the given nested key path `{key}` is already set to a list'
```

Comment:
`is already set to a list` might be true if you are trying to assign a value to a key that is currently a list, but this would also trigger if the sub path already has a value assigned to it, wouldn't it? Let's say the path `some.path` has a value assigned to it, then trying to assign a value to `some.path.namespace` will also raise this `TypeError` and have a misleading error message. Would probably be useful to distinguish these cases individually.

Reply:
Actually there are not too many options here if I understood it correctly.
But you're probably right that we have to move the `isinstance(Awaitable, ...)` check here, otherwise we might create a subdict inside an `Awaitable` (what I prevented below for the the leave-node, but may happen inside the loop already).
But I'll take another look.

Reply:
ugh, okay, got it: once an Awaitable is resolved it can be many different values.